<h2 align=center><font size = 5>Applied Data Science Capstone</font></h2>

#### Research Topic: Identifying the most Ideal location in Toronto to set up a restaurant

In [1]:
# Import Libraries
import requests #library for data analsysis

import lxml.html as lh
import bs4 as bs
import urllib.request

import numpy as np # library to handle data in a vectorized manner
import pandas as pd # library for data analsysis

In [4]:
## Source URL for post codes
url   = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"

In [5]:
# BS4, passed to scrape_table_bs4
# create colums <tableClassName> <expected numberOfColumns>

def scrape_table_bs4(cname,cols):
    page  = urllib.request.urlopen(url).read()
    soup  = bs.BeautifulSoup(page,'lxml')
    table = soup.find("table",class_=cname)
    header = [head.findAll(text=True)[0].strip() for head in table.find_all("th")]
    data   = [[td.findAll(text=True)[0].strip() for td in tr.find_all("td")]
              for tr in table.find_all("tr")]
    data    = [row for row in data if len(row) == cols]
    # Store data to this temporary dataframe
    raw_df = pd.DataFrame(data,columns=header)
    return raw_df

# Parsing using xpath.
def scrape_table_lxml(XPATH,cols):
    page = requests.get(url)
    doc = lh.fromstring(page.content)
    table_content = doc.xpath(XPATH)
    for table in table_content:
        headers = [th.text_content().strip() for th in table.xpath('//th')]
        headers = headers[0:3]
        data    = [[td.text_content().strip() for td in tr.xpath('td')] 
                   for tr in table.xpath('//tbody/tr')]
        data    = [row for row in data if len(row) == cols]
        raw_df = pd.DataFrame(data,columns=headers)
        return raw_df

In [6]:
#Call functional scope test in beautifulSoup
raw_TorontoPostalCodes = scrape_table_bs4("wikitable",3)

#Test in lxml ( for xpath based extraction)
#raw_TorontoPostalCodes = scrape_table_lxml("/html/body/div[3]/div[3]/div[4]/div/table[1]",3)

print("# Toronto Postal codes stored in data")
print(raw_TorontoPostalCodes.info(verbose=True))

# Toronto Postal codes stored in data
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180 entries, 0 to 179
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Postal Code    180 non-null    object
 1   Borough        180 non-null    object
 2   Neighbourhood  180 non-null    object
dtypes: object(3)
memory usage: 4.3+ KB
None


#####  Data Prepocessing and Transformation: Scraped data table from wiki contains some un-wanted entries and needs some cleaning up. 

In [7]:
# Focus only on those cell that had an assigned borough. ignore else. 
TorontoPostalCodes=raw_TorontoPostalCodes[~raw_TorontoPostalCodes['Borough'].isin(['Not assigned'])]

# Sort and Reset index in the table.
TorontoPostalCodes=TorontoPostalCodes.sort_values(by=['Postal Code','Borough','Neighbourhood'], ascending=[1,1,1]).reset_index(drop=True)

# *If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.

TorontoPostalCodes.loc[TorontoPostalCodes['Neighbourhood'] == 'Not assigned', ['Neighbourhood']] = TorontoPostalCodes['Borough']
check_unassigned_post_state_sample = TorontoPostalCodes.loc[TorontoPostalCodes['Borough'] == 'Queen\'s Park']
#print('DEBUG:',check_unassigned_post_state_sample) ; # Print sample borough problem post state

# More than one neighborhood can exist in one postal code area. 
# For example, in the table on the Wikipedia page, you will notice that M5A is listed twice 
# and has two neighborhoods: Harbourfront and Regent Park. 
# These two rows will be combined into one row with the neighborhoods separated with a comma.

TorontoPostalCodes = TorontoPostalCodes.groupby(['Postal Code','Borough'])['Neighbourhood'].apply(', '.join).reset_index()

In [8]:
TorontoPostalCodes.head() # Quick check at DF

Postal Code      Borough                           Neighbourhood
0         M1B  Scarborough                          Malvern, Rouge
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2         M1E  Scarborough       Guildwood, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                               Cedarbrae

Foursquare API to explore neighborhoods in selected cities in Toronto, in the hope of sorting, clustering and modeling the attributes of one location relative to the other. The Foursquare explore function will be used to get the most common venue categories in each neighborhood, and then use this feature to group the neighborhoods into clusters. 

In [11]:
# Importing libraries for mapping and K Means
#!conda install -c conda-forge geopy --yes        # if needed
#!conda install -c conda-forge folium=0.5.0 --yes # if needed

import numpy as np # library to handle data in a vectorized manner
import pandas as pd # library for data analsysis

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files


from geopy.geocoders import Nominatim 
GeoLocator = Nominatim(user_agent='My-IBMNotebook')# convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans


import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [12]:
# backing source data of postal codes, borough + neighbors + Longitude and Latitude
toronto_task1_csv = "Toronto.TASK_1_df.csv" # Contains postcodes and names
toronto_task2_csv = "Toronto.TASK_II_df.csv" # Contains postcodes and LAT LONG

In [13]:
toronto_neighborhoods = pd.read_csv(toronto_task2_csv)

In [14]:
toronto_neighborhoods.shape # take a look at the dimentional features of the data
toronto_neighborhoods.head(5)

Borough                           Neighbourhood   Latitude  Longitude
0  Scarborough                          Malvern, Rouge  43.806686 -79.194353
1  Scarborough  Highland Creek, Port Union, Rouge Hill  43.784535 -79.160497
2  Scarborough       Guildwood, Morningside, West Hill  43.763573 -79.188711
3  Scarborough                                  Woburn  43.770992 -79.216917
4  Scarborough                               Cedarbrae  43.773136 -79.239476

Use geopy library to get the latitude and longitude values of Toronto Canada.¶

In [15]:
address = 'Toronto, Ontario Canada'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto Canada are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto Canada are 43.6534817, -79.3839347.


In [19]:
# create map of Toronto using latitude and longitude values, with neighborhoods superimposed
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_neighborhoods['Latitude'], toronto_neighborhoods['Longitude'], toronto_neighborhoods['Borough'], toronto_neighborhoods['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=6,
        popup=label,
        color='red',
        fill=True,
        fill_color='green',
        fill_opacity=0.5,
        parse_html=False).add_to(map_toronto)

    map_toronto

For this task, our target analysis to Neighbohoods in East, West and Central Toronto, but subsetting dataframe, i.e, we look at just a part, boroughs containing word Toronto.

In [18]:
toronto_data = toronto_neighborhoods[toronto_neighborhoods['Borough'].str.contains("Toronto")].reset_index(drop=True)
print(toronto_data.shape)
toronto_data.head(10)

(38, 4)


Borough                                      Neighbourhood  \
0     East Toronto                                        The Beaches   
1     East Toronto                       Riverdale, The Danforth West   
2     East Toronto                     India Bazaar, The Beaches West   
3     East Toronto                                    Studio District   
4  Central Toronto                                      Lawrence Park   
5  Central Toronto                                   Davisville North   
6  Central Toronto                                 North Toronto West   
7  Central Toronto                                         Davisville   
8  Central Toronto                        Moore Park, Summerhill East   
9  Central Toronto  Deer Park, Forest Hill SE, Rathnelly, South Hi...   

    Latitude  Longitude  
0  43.676357 -79.293031  
1  43.679557 -79.352188  
2  43.668999 -79.315572  
3  43.659526 -79.340923  
4  43.728020 -79.388790  
5  43.712751 -79.390197  
6  43.715383 -79.405678  
7  43.704324 -79.388790  
8  43.689574 -79.383160  
9  43.686412 -79.400049

In [20]:
# Now, redo the Map of Toronto, but here focus on the area of research, by using the same coordinates for the previous view, take a closer look at Central Toronto
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=9,
        popup=label,
        color='orange',
        fill=True,
        fill_color='yellow',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Utilizing the Foursquare API to explore and segment neighborhoods. This will provide us with a more contextaul view of the city's economic centers.

#### Methodology: 
##### At this stage we general outline a step by step analytical walk through on the overal framework. It provides for a discussion and description of exploratory data analysis, inferential statistical testing conducted, and other methodological approaches. The k-means clustering algorithm will be used for the analysis. Finally, use the Folium library to visualize the neighborhoods in Toronto and their emerging clusters.

In [21]:
task1_data_csv = "Toronto.TASK_1_df.csv"
TorontoPostalCodes = pd.read_csv(task1_data_csv).set_index("Postcode")
TorontoPostalCodes.rename_axis("Postal Code", axis='index', inplace=True)
TorontoPostalCodes.head(5)

Borough                           Neighbourhood
Postal Code                                                     
M1B          Scarborough                          Malvern, Rouge
M1C          Scarborough  Highland Creek, Port Union, Rouge Hill
M1E          Scarborough       Guildwood, Morningside, West Hill
M1G          Scarborough                                  Woburn
M1H          Scarborough                               Cedarbrae

##### Retrieve and transform postal code coordinate values list.

Now that we have built a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name, in order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood. We will use the Geocoder Python package to do so: https://geocoder.readthedocs.io/index.html.

In [22]:
toronto_geocsv = 'https://cocl.us/Geospatial_data'
!wget -q -O 'toronto_m.geospatial_data.csv' toronto_geocsv
geocsv_data = pd.read_csv(toronto_geocsv).set_index("Postal Code")
geocsv_data.head(5)

'wget' is not recognized as an internal or external command,
operable program or batch file.


Latitude  Longitude
Postal Code                      
M1B          43.806686 -79.194353
M1C          43.784535 -79.160497
M1E          43.763573 -79.188711
M1G          43.770992 -79.216917
M1H          43.773136 -79.239476

In [23]:
# combining Two data frames with mapped postcodes
toronto_neighborhoods = TorontoPostalCodes.join(geocsv_data)
toronto_neighborhoods.head(5)

Borough                           Neighbourhood   Latitude  \
Postal Code                                                                   
M1B          Scarborough                          Malvern, Rouge  43.806686   
M1C          Scarborough  Highland Creek, Port Union, Rouge Hill  43.784535   
M1E          Scarborough       Guildwood, Morningside, West Hill  43.763573   
M1G          Scarborough                                  Woburn  43.770992   
M1H          Scarborough                               Cedarbrae  43.773136   

             Longitude  
Postal Code             
M1B         -79.194353  
M1C         -79.160497  
M1E         -79.188711  
M1G         -79.216917  
M1H         -79.239476

In [24]:
# always look at newly creat dataframes 
toronto_neighborhoods.to_csv('Toronto.TASK_II_df.csv',index=False)

In [25]:
toronto_neighborhoods.shape

(103, 4)

In [26]:
toronto_neighborhoods.dtypes

Borough           object
Neighbourhood     object
Latitude         float64
Longitude        float64
dtype: object

In [27]:
toronto_neighborhoods.describe()

Latitude   Longitude
count  103.000000  103.000000
mean    43.704608  -79.397153
std      0.052463    0.097146
min     43.602414  -79.615819
25%     43.660567  -79.464763
50%     43.696948  -79.388790
75%     43.745320  -79.340923
max     43.836125  -79.160497

In [28]:
# Take note to not disclose private or sensitive personal information. Here, I pass hidden a paramenter to redact my credentials

CLIENT_ID ='VWLF3XWZMCTBA05OGRE5CCBPMCCLF3H1CCK53S5IU3D3TN3Y' # portion hidden from view' # your Foursquare ID
CLIENT_SECRET = 'N34CA1DLJEVQWFIRIOCKULTJMHLAVANGZ0CWTZDGEEKVOFDS'# portion hidden from view' # your Foursquare Secret
VERSION = '20200101'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: VWLF3XWZMCTBA05OGRE5CCBPMCCLF3H1CCK53S5IU3D3TN3Y
CLIENT_SECRET:N34CA1DLJEVQWFIRIOCKULTJMHLAVANGZ0CWTZDGEEKVOFDS


##### We begin by exploring Neighbourhood in Toronto, using the following foursquare api query url, search venues on all boroughs in selected Toronto neighborhoods
###### https://api.foursquare.com/v2/venues/search?client_id=CLIENT_ID&client_secret=CLIENT_SECRET&ll=LATITUDE,LONGITUDE&v=VERSION&qu

In [29]:
# The following function retrieves the venues given the names and coordinates and stores it into dataframe.
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)
# The following function retrieves the venues given the names and coordinates and stores it into dataframe.
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [30]:
# Retrieve all venues given the Addresses
toronto_neighborhoods = toronto_data
toronto_venues = getNearbyVenues(names=toronto_neighborhoods['Neighbourhood'],
                                   latitudes=toronto_neighborhoods['Latitude'],
                                   longitudes=toronto_neighborhoods['Longitude']
                                  )

The Beaches
Riverdale, The Danforth West
India Bazaar, The Beaches West
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park, Summerhill East
Deer Park, Forest Hill SE, Rathnelly, South Hill, Summerhill West
Rosedale
Cabbagetown, St. James Town
Church and Wellesley
Harbourfront, Regent Park
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Adelaide, King, Richmond
Harbourfront East, Toronto Islands, Union Station
Design Exchange, Toronto Dominion Centre
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North, Forest Hill West
North Midtown, The Annex, Yorkville
Harbord, University of Toronto
Chinatown, Grange Park, Kensington Market
Bathurst Quay, CN Tower, Harbourfront West, Island airport, King and Spadina, Railway Lands, South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place, Underground city
Christie
Dovercourt Village, Dufferin
Little Portugal, Trinity
Brockton, Exhibition Place, Parkdale Village
High Park, The 

##### Checking size of resulting dataframe obtained above

In [31]:
print(toronto_venues.shape)
toronto_venues.head(5)

(831, 7)


Neighbourhood  Neighborhood Latitude  \
0                   The Beaches              43.676357   
1                   The Beaches              43.676357   
2                   The Beaches              43.676357   
3                   The Beaches              43.676357   
4  Riverdale, The Danforth West              43.679557   

   Neighborhood Longitude                               Venue  Venue Latitude  \
0              -79.293031                   Glen Manor Ravine       43.676821   
1              -79.293031  The Big Carrot Natural Food Market       43.678879   
2              -79.293031                 Grover Pub and Grub       43.679181   
3              -79.293031                       Upper Beaches       43.680563   
4              -79.352188                       MenEssentials       43.677820   

   Venue Longitude     Venue Category  
0       -79.293942              Trail  
1       -79.297734  Health Food Store  
2       -79.297215                Pub  
3       -79.292869       Neighborhood  
4       -79.351265     Cosmetics Shop

In [32]:
# Count of venues were returned for each Borough
toronto_venues.groupby('Neighbourhood').count()

Neighborhood Latitude  \
Neighbourhood                                                               
Adelaide, King, Richmond                                               30   
Bathurst Quay, CN Tower, Harbourfront West, Isl...                     16   
Berczy Park                                                            30   
Brockton, Exhibition Place, Parkdale Village                           22   
Business Reply Mail Processing Centre 969 Eastern                      18   
Cabbagetown, St. James Town                                            30   
Central Bay Street                                                     30   
Chinatown, Grange Park, Kensington Market                              30   
Christie                                                               17   
Church and Wellesley                                                   30   
Commerce Court, Victoria Hotel                                         30   
Davisville                                                             30   
Davisville North                                                        8   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                     17   
Design Exchange, Toronto Dominion Centre                               30   
Dovercourt Village, Dufferin                                           14   
First Canadian Place, Underground city                                 30   
Forest Hill North, Forest Hill West                                     4   
Garden District, Ryerson                                               30   
Harbord, University of Toronto                                         30   
Harbourfront East, Toronto Islands, Union Station                      30   
Harbourfront, Regent Park                                              30   
High Park, The Junction South                                          25   
India Bazaar, The Beaches West                                         19   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               30   
Moore Park, Summerhill East                                             2   
North Midtown, The Annex, Yorkville                                    20   
North Toronto West                                                     23   
Parkdale, Roncesvalles                                                 14   
Riverdale, The Danforth West                                           30   
Rosedale                                                                4   
Roselawn                                                                1   
Runnymede, Swansea                                                     30   
St. James Town                                                         30   
Stn A PO Boxes 25 The Esplanade                                        30   
Studio District                                                        30   
The Beaches                                                             4   

                                                    Neighborhood Longitude  \
Neighbourhood                                                                
Adelaide, King, Richmond                                                30   
Bathurst Quay, CN Tower, Harbourfront West, Isl...                      16   
Berczy Park                                                             30   
Brockton, Exhibition Place, Parkdale Village                            22   
Business Reply Mail Processing Centre 969 Eastern                       18   
Cabbagetown, St. James Town                                             30   
Central Bay Street                                                      30   
Chinatown, Grange Park, Kensington Market                               30   
Christie                                                                17   
Church and Wellesley                                                    30   
Commerce Court, Victoria Hotel     

##### Identfying the number of unique categories that be found among all the returned venues

In [33]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 190 uniques categories.


In [34]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Airport  Airport Food Court  Airport Gate  \
0                   The Beaches        0                   0             0   
1                   The Beaches        0                   0             0   
2                   The Beaches        0                   0             0   
3                   The Beaches        0                   0             0   
4  Riverdale, The Danforth West        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0                 0                   0              0          0           0   
1                 0                   0              0          0           0   
2                 0                   0              0          0           0   
3                 0                   0              0          0           0   
4                 0                   0              0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Basketball Stadium  Beer Bar  Beer Store  \
0           0       0     0    0                   0         0           0   
1           0       0     0    0                   0         0           0   
2           0       0     0    0                   0         0           0   
3           0       0     0    0                   0         0           0   
4           0       0     0    0                   0         0           0   

   Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                   0       0              0          0         0   
1                   0       0              0          0         0   
2                   0       0              0          0         0   
3                   0       0              0          0         0   
4                   0       0              0          0         0   

   Breakfast Spot  Brewery  Bubble Tea Shop  Burger Joint  Burrito Place  \
0               0        0                0             0              0   
1               0        0                0             0              0   
2               0        0                0             0              0   
3               0        0                0             0              0   
4               0        0                0             0              0   

   Bus Line  Butcher  Café  Cajun / Creole Restaurant  Camera Store  \
0         0        0     0                          0             0   
1         0        0     0                          0             0   
2         0        0     0                          0             0   
3         0        0     0                          0             0   
4         0        0     0                          0             0   

   Candy Store  Caribbean Restaurant  Cheese Shop  Chinese Restaurant  \
0            0                     0            0                   0   
1            0                     0            0                   0   
2            0                     0            0                   0   
3            0                     0            0                   0   
4            0 

In [35]:
toronto_onehot.shape

(831, 191)

###### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [37]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Airport  \
0                            Adelaide, King, Richmond   0.0000   
1   Bathurst Quay, CN Tower, Harbourfront West, Is...   0.0625   
2                                         Berczy Park   0.0000   
3        Brockton, Exhibition Place, Parkdale Village   0.0000   
4   Business Reply Mail Processing Centre 969 Eastern   0.0000   
5                         Cabbagetown, St. James Town   0.0000   
6                                  Central Bay Street   0.0000   
7           Chinatown, Grange Park, Kensington Market   0.0000   
8                                            Christie   0.0000   
9                                Church and Wellesley   0.0000   
10                     Commerce Court, Victoria Hotel   0.0000   
11                                         Davisville   0.0000   
12                                   Davisville North   0.0000   
13  Deer Park, Forest Hill SE, Rathnelly, South Hi...   0.0000   
14           Design Exchange, Toronto Dominion Centre   0.0000   
15                       Dovercourt Village, Dufferin   0.0000   
16             First Canadian Place, Underground city   0.0000   
17                Forest Hill North, Forest Hill West   0.0000   
18                           Garden District, Ryerson   0.0000   
19                     Harbord, University of Toronto   0.0000   
20  Harbourfront East, Toronto Islands, Union Station   0.0000   
21                          Harbourfront, Regent Park   0.0000   
22                      High Park, The Junction South   0.0000   
23                     India Bazaar, The Beaches West   0.0000   
24                                      Lawrence Park   0.0000   
25                           Little Portugal, Trinity   0.0000   
26                        Moore Park, Summerhill East   0.0000   
27                North Midtown, The Annex, Yorkville   0.0000   
28                                 North Toronto West   0.0000   
29                             Parkdale, Roncesvalles   0.0000   
30                       Riverdale, The Danforth West   0.0000   
31                                           Rosedale   0.0000   
32                                           Roselawn   0.0000   
33                                 Runnymede, Swansea   0.0000   
34                                     St. James Town   0.0000   
35                    Stn A PO Boxes 25 The Esplanade   0.0000   
36                                    Studio District   0.0000   
37                                        The Beaches   0.0000   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0               0.0000        0.0000           0.000           0.0000   
1               0.0625        0.0625           0.125           0.1875   
2               0.0000        0.0000           0.000           0.0000   
3               0.0000        0.0000           0.000           0.0000   
4               0.0000        0.0000           0.000           0.0000   
5               0.0000        0.0000           0.000           0.0000   
6               0.0000        0.0000           0.000           0.0000   
7               0.0000        0.0000           0.000           0.0000   
8               0.0000        0.0000           0.000           0.0000   
9               0.0000        0.0000           0.000           0.0000   
10              0.0000        0.0000           0.000           0.0000   
11              0.0000        0.0000           0.000           0.0000   
12              0.0000        0.0000           0.000           0.0000   
13              0.0000        0.0000           0.000           0.0000   
14              0.0000        0.0000           0.000           0.0000   
15              0.0000        0.0000           0.000           0.0000   
16              0.0000        0.0000           0.000           0.0000   
17              0.0000        0.0000           0.000           0.0000   
18              0.0000        0.0000           0.000           0.0000   
19   

#### 4. Analysis and Results on Each Borough Neighborhood:

In [38]:
#Check new size:
toronto_grouped.shape

(38, 191)

In [39]:
# Let's put that into a pandas dataframe
num_top_venues = 5
for neigh in toronto_grouped['Neighbourhood']:
    print("----"+neigh+"----")
    temp = toronto_grouped[toronto_grouped['Neighbourhood'] == neigh].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
         venue  freq
0         Café  0.10
1  Coffee Shop  0.10
2   Steakhouse  0.07
3        Hotel  0.07
4        Plaza  0.03


----Bathurst Quay, CN Tower, Harbourfront West, Island airport, King and Spadina, Railway Lands, South Niagara----
                 venue  freq
0      Airport Service  0.19
1       Airport Lounge  0.12
2              Airport  0.06
3   Airport Food Court  0.06
4  Rental Car Location  0.06


----Berczy Park----
                venue  freq
0         Coffee Shop  0.10
1  Seafood Restaurant  0.07
2        Cocktail Bar  0.07
3      Farmers Market  0.07
4         Fish Market  0.03


----Brockton, Exhibition Place, Parkdale Village----
                venue  freq
0                Café  0.14
1         Coffee Shop  0.09
2      Breakfast Spot  0.09
3       Grocery Store  0.05
4  Italian Restaurant  0.05


----Business Reply Mail Processing Centre 969 Eastern----
                venue  freq
0  Light Rail Station  0.11
1         Yoga Studio

##### First, pass a function to Python to sort the locations in descending order, i.e A-Z.

In [40]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [41]:
# create columns according to number of top venues
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.shape

(38, 11)

#### Observations 

##### I then move to cluster the Neighborhoods, into groups based on the frequency and attributes using K-means Clustering Algorithim. k-means clustering aims to partition n observations into k clusters in which each observation belongs to the cluster with the nearest mean, serving as a prototype of the cluster.

In [66]:
# Run k-means to cluster the neighborhood into 3 clusters 
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)
# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=1).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
print(kmeans.labels_[0:20])
print(len(kmeans.labels_))

[2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2]
38


In [67]:
#Let's take a look at our new dataframe again
toronto_neighborhoods.shape

(38, 5)

##### We now have a framework through which we can make some determinations, that is, find the best place to locate a new business venture, for example, a resturant across Toronto

In [68]:
toronto_neighborhoods.head()

Borough                   Neighbourhood   Latitude  Longitude  \
0     East Toronto                     The Beaches  43.676357 -79.293031   
1     East Toronto    Riverdale, The Danforth West  43.679557 -79.352188   
2     East Toronto  India Bazaar, The Beaches West  43.668999 -79.315572   
3     East Toronto                 Studio District  43.659526 -79.340923   
4  Central Toronto                   Lawrence Park  43.728020 -79.388790   

   Cluster Labels  
0               3  
1               3  
2               3  
3               3  
4               3

In [69]:
toronto_merged = toronto_neighborhoods

# add clustering labels
toronto_merged['Cluster Labels'] = kmeans.labels_

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged  # check the last columns!

Borough                                      Neighbourhood  \
0       East Toronto                                        The Beaches   
1       East Toronto                       Riverdale, The Danforth West   
2       East Toronto                     India Bazaar, The Beaches West   
3       East Toronto                                    Studio District   
4    Central Toronto                                      Lawrence Park   
5    Central Toronto                                   Davisville North   
6    Central Toronto                                 North Toronto West   
7    Central Toronto                                         Davisville   
8    Central Toronto                        Moore Park, Summerhill East   
9    Central Toronto  Deer Park, Forest Hill SE, Rathnelly, South Hi...   
10  Downtown Toronto                                           Rosedale   
11  Downtown Toronto                        Cabbagetown, St. James Town   
12  Downtown Toronto                               Church and Wellesley   
13  Downtown Toronto                          Harbourfront, Regent Park   
14  Downtown Toronto                           Garden District, Ryerson   
15  Downtown Toronto                                     St. James Town   
16  Downtown Toronto                                        Berczy Park   
17  Downtown Toronto                                 Central Bay Street   
18  Downtown Toronto                           Adelaide, King, Richmond   
19  Downtown Toronto  Harbourfront East, Toronto Islands, Union Station   
20  Downtown Toronto           Design Exchange, Toronto Dominion Centre   
21  Downtown Toronto                     Commerce Court, Victoria Hotel   
22   Central Toronto                                           Roselawn   
23   Central Toronto                Forest Hill North, Forest Hill West   
24   Central Toronto                North Midtown, The Annex, Yorkville   
25  Downtown Toronto                     Harbord, University of Toronto   
26  Downtown Toronto          Chinatown, Grange Park, Kensington Market   
27  Downtown Toronto  Bathurst Quay, CN Tower, Harbourfront West, Is...   
28  Downtown Toronto                    Stn A PO Boxes 25 The Esplanade   
29  Downtown Toronto             First Canadian Place, Underground city   
30  Downtown Toronto                                           Christie   
31      West Toronto                       Dovercourt Village, Dufferin   
32      West Toronto                           Little Portugal, Trinity   
33      West Toronto       Brockton, Exhibition Place, Parkdale Village   
34      West Toronto                      High Park, The Junction South   
35      West Toronto                             Parkdale, Roncesvalles   
36      West Toronto                                 Runnymede, Swansea   
37      East Toronto  Business Reply Mail Processing Centre 969 Eastern   

     Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0   43.676357 -79.293031               2          Neighborhood   
1   43.679557 -79.352188               2      Greek Restaurant   
2   43.668999 -79.315572               2        Sandwich Place   
3   43.659526 -79.340923               2                  Café   
4   43.728020 -79.388790               2                  Park   
5   43.712751 -79.390197               2                  Park   
6   43.715383 -79.405678               2        Clothing Store   
7   43.704324 -79.388790               2          Dessert Shop   
8   43.689574 -79.383160               2                   Gym   
9   43.686412 -79.400049               2    Light Rail Station   
10  43.679563 -79.377529               2                  Park   
11  43.667967 -79.367675               2    Italian Restaurant   
12  43.665860 -79.383160               2           Coffee Shop   
13  43.654260 -79.360636               2           Coffee Shop   
14  43.657162 -79.378937               2                  Café   
15  43.651494 -79.375418       

#### Conclusion and Recommendations:

#### Now, we can finally do visualization of the clusters we produced above. By evaluation the clusters of the varios locatioins, we can better under the communities and dertermine where to locate.

In [70]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'],kmeans.labels_):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.5).add_to(map_clusters)
       
map_clusters

#### Conclusion: <br>
After examining the clusters,  I believe and would recommend that "Business Reply Mail Processing Centre 969 Eastern center" would be a good place to start a takeaway foodshop as one could expect many walking due to presnse of light rail. One could also note that there are no takeaway/resturants listed in the top 5 locations at this address. 